---
title: "Yeast Data Preparation: Filtering, Aggregation, and Normalisation"
author: "M.Soria"
date: "`r Sys.Date()`"
output:
  html_document:
  toc: true
theme: cosmo
---

### Introduction

The following code consists of the data preparation steps for the RNA-Seq data for Saccharomyces cerevisiae. 
The data preparation consists of quality filtering, aggregation, and normalisation.

### Prerequisite libraries

```{r Libraries, echo=TRUE}

suppressPackageStartupMessages({
    library(R.utils)
    library(dplyr)
    library(tidyr)
    library(data.table)
    library(RColorBrewer)
    library(Polychrome)
    library(gplots)
    library(edgeR)
    library(InterMineR)
})

# Special Installation for edgeR

# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("edgeR")

source("functions.R")

```

### 1. Download and unzip the Yeast data 

```{r Download Data, echo=TRUE}

# put in an if statement to check if object exists

if (!file.exists("Data/scerevisiae_se.tsv.bz2")){
    download.file("http://dee2.io/mx/scerevisiae_se.tsv.bz2", 
              destfile = "Data/scerevisiae_se.tsv.bz2")}

# fread error msg: This file is very unusual: it ends abruptly without a final newline, 
# and also its size is a multiple of 4096 bytes.
# Please properly end the last row with a newline using for example 'echo >> file' to avoid this time to copy. 

# needed to go back to bunzip2
if (!file.exists("Data/scerevisiae_se.tsv")){
    bunzip2("Data/scerevisiae_se.tsv.bz2", destname="Data/scerevisiae_se.tsv")}

scerevisiae_data <- fread("Data/scerevisiae_se.tsv", sep = '\t', header = FALSE)

head(scerevisiae_data)

```

### 2. Download, read, and unzip the QC data

```{r QC Data, echo=TRUE}

if (!file.exists("Data/scerevisiae_qc.tsv.bz2")){
    download.file("http://dee2.io/mx/scerevisiae_qc.tsv.bz2", 
              destfile = "Data/scerevisiae_qc.tsv.bz2")}

quality_data <- fread("Data/scerevisiae_qc.tsv.bz2", sep = '\t', 
  col.names = c("Database", "Description", "Details"))

head(quality_data)

```

### 3. Filter and quantify databases according to QC_SUMMARY = PASS, WARN, or FAIL.  

```{r QC Data Summary, echo=TRUE}

# count the total number of QC_SUMMARY entries for checking
qc_summary <- quality_data[quality_data$Description == 'QC_SUMMARY',]
total_summary <- nrow(qc_summary)

# count the total number of databases marked as PASS
qc_pass <- filter(quality_data, grepl("PASS", Details))
total_pass <- nrow(qc_pass)

# count the total number of databases marked as WARN
qc_warn <- filter(quality_data, grepl("WARN", Details))
total_warn <- nrow(qc_warn)

# count the total number of databases marked as FAIL
qc_fail <- filter(quality_data, grepl("FAIL", Details))
total_fail <- nrow(qc_fail)

# Summary of counts. Addition of each PASS, WARN, and FAIL counts equals to total_summary.
quality_summary <- data.frame(total_pass, total_warn, total_fail, total_summary)
quality_summary

```

### 4. Select all the databses from scerevisiae_se.tsv with QC_SUMMARY == 'PASS' 

```{r Filter yeast bulk data, echo=TRUE}

# filter all the sample names with QC_SUMMARY == 'PASS' from step 3
databse_pass <- filter(quality_data, grepl("PASS", Details)) 

# convert the rows into string and store on a list to use for filtering Yeast data
databse_pass_list <- as.list(as.character(databse_pass$Database))

# Filter Yeast data using generated list
scerevisiae_pass <- scerevisiae_data[scerevisiae_data$V1 %in% databse_pass_list,] 

# Convert format from long to wide
scerevisiae_pass_wide <- scerevisiae_pass %>% pivot_wider(names_from = "V1", values_from = "V3")

# Convert tibble to data frame and assign column 1 as rowname
scerevisiae_pass_wide <- as.data.frame(scerevisiae_pass_wide)
rownames(scerevisiae_pass_wide) <- scerevisiae_pass_wide[,1]
scerevisiae_pass_wide <- scerevisiae_pass_wide[,-1]

scerevisiae_pass_wide[1:10,1:6]

```

### 5. Aggregate multiple runs, if any, (SRR) to its corresponding experiment (SRX) with PASS only databases

```{r SRR to SRX Aggregation, echo=TRUE}

# Download and read Metadata summary

if (!file.exists("Data/scerevisiae_metadata.tsv.cut")){
    download.file("http://dee2.io/metadata/scerevisiae_metadata.tsv.cut", 
              destfile = "Data/scerevisiae_metadata.tsv.cut")}
    
yeast_metadata <- read.csv("Data/scerevisiae_metadata.tsv.cut", sep = '\t')

# Filter metadata to include only "passed" samples
yeast_metadata_pass <- yeast_metadata[yeast_metadata$SRR_accession %in% databse_pass_list,]

# Assign column 1 as rownames 
rownames(yeast_metadata_pass) <- yeast_metadata_pass[,1]
yeast_metadata_pass <- yeast_metadata_pass[,-1]

# Put the filtered yeast GeneCount data and the filtered metadata dataframes into a list
yeast_count_metadata <- list(GeneCounts = scerevisiae_pass_wide, MetadataSummary = yeast_metadata_pass)

# Apply both dataframes to the function
agg <- srx_agg(yeast_count_metadata)

# Check for samples with zero total counts
colSums(agg) == 0

# Histogram showing samples (columns) counts
hist(colSums(agg), breaks = 30)

# Filtering samples with more than 1 million reads
agg_f_col <- agg[, which(colSums(agg) >= 1e6)]

# Filtering genes with more than zero counts 
agg_f_row <- agg[which(rowSums(agg) > 0), ]

# No difference between the filtered and unfiltered columns  
dim(agg_f_col)
dim(agg)

# Rows ommited for genes with more than zero counts
dim(agg_f_row)

# Save agg object
saveRDS(agg, "agg.rds")

```

### 6. Normalisation of Data (with PASS only databases)

```{r }

# NORMALISATION FOR LIBRARY SIZE BIAS (cpm)
myCPM <- cpm(agg)
# add a small number to the entire dataframe to account for zero reads
myCPM <- myCPM + 0.01
# Filtering to remove lowly expressed genes
thresh <- myCPM > 0.6
# check if there are zero values. Result should be FALSE.
all(apply(apply(myCPM, 2, function(x) x==0), 2, any))

# Keep genes that have at least 2 TRUEs in each row
keep <- rowSums(thresh) >= 2

# Subset the rows of data to keep the more highly expressed genes
counts_keep <- myCPM[keep,]
summary(keep)
dim(counts_keep)

# A CPM of 0.6 is used as it corresponds to a count of 10-15 for the library sizes in this data set.
# If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample.
# As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.6.
# You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.
# Note: When in doubt, a threshold of 1 CPM in at least minimum group sample size is a good rule of thumb.

# Check whether our threshold of 0.6 does indeed correspond to a count of about 10-15
# Add a vertical line at 0.6 CPM, and a horizontal line at Counts = 10
plot(myCPM[,1], agg[,1], ylim=c(0,50), xlim=c(0,3), abline(v=0.6, h=10, col="blue"), 
  ylab = "Counts", xlab = "CPM", main = "Counts-per-million (CPM) vs Counts")

# Convert counts to DGEList object (from edgeR library)
dgeObj <- DGEList(counts_keep)

#NORMALISATION FOR COMPOSITION BIAS (TMM normalisation)
# Apply normalisation to DGEList object
dgeObj <- calcNormFactors(dgeObj) 
# dgeObj$samples

# Get log2 counts per million
logcounts_PASS <- cpm(dgeObj,log=TRUE)
saveRDS(logcounts_PASS, "logcounts_PASS.rds")

# Check distributions of samples
hist(logcounts_PASS)
hist(counts_keep)

all(apply(apply(logcounts_PASS, 2, function(x) x==0), 2, any))

```

### 7. Heirarchical Clustering with PASS only databases

```{r echo=TRUE,fig.height=8,fig.width=8}

# This chunk can be reused for the PASS + Warn or the WARN filters by changing the content of the filename object. 
# Do not forget to change the name of the variable where you will save the clusters for different inputs. 

# Save the logcounts_PASS.rds object from the data preparation step (YeastDataPrep.Rmd)
filename <- "logcounts_PASS.rds"
# Read the logcounts_PASS.rds object and use it in the code
logcounts <- readRDS(filename)

# Hierarchical Clustering
# set seed = 42 for reproducibility of clustering

set.seed(42)
cl <-as.dist(1-cor(t(logcounts), method="spearman"))
hr <- hclust(cl , method="complete")
saveRDS(hr, "hr_PASS_PandC.rds")

# optimizing the cluster size
# 1.605 was determined by trial and error to get 100 total 
# clusters as a starting point of analysis

mycl <- cutree(hr, h=max(hr$height/1.605))
mycl_length <- length(unique(mycl))

# Check the number of clusters. Can be adjusted by changing the h=max denominator
mycl_length

# Save the clusters value 
saveRDS(mycl, "mycl_PASS_PandC.rds")

# Prepare Cluster data frame
clusters <- as.data.frame(mycl)
colnames(clusters) <- "ClusterNumber"

# make a new column, GeneID, from rownames
clusters$GeneID <- rownames(clusters)

# Save clusters object as rds
saveRDS(clusters, "clusters_PandC.rds")

# # Visualizations
# if (!exists("P100")){
#   P100 <- createPalette(100,  c("#ff0000", "#00ff00", "#0000ff"))}
# 
# clusterCols <- P100[1:mycl_length]
# myClusterSideBar <- clusterCols[mycl]
# colfunc <- colorRampPalette(c("blue", "white", "red"))
# if (!file.exists("Data/scerevisiae_pass.txt")){
#   write.table(mycl,file="Data/scerevisiae_pass.txt",quote=F,sep="\t")}
# 
# 
# start.time <- Sys.time()
# # create a 5 x 5 inch png image
# png("Data/yeast_heatmaps_PASS_normalization_saleNone.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)
# 
# # create the heatmap
# heatmap.2(logcounts, main="S.cerevisiae PASS Samples",  Rowv=as.dendrogram(hr),
#           dendrogram="both", scale="none", col = colfunc(25), trace="none",
#           RowSideColors= myClusterSideBar, margins = c(5,5))
# 
# dev.off()
# 
# end.time <- Sys.time()
# time.taken <- end.time - start.time
# time.taken
# 
# 
# start.time <- Sys.time()
# # create a 5 x 5 inch png image
# png("Data/yeast_heatmaps_PASS_normalization_saleRow.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)
# 
# # create the heatmap
# heatmap.2(logcounts, main="S.cerevisiae PASS Samples",  Rowv=as.dendrogram(hr),
#           dendrogram="both", scale="row", col = colfunc(25), trace="none",
#           RowSideColors= myClusterSideBar, margins = c(5,5))
# 
# dev.off()
# 
# end.time <- Sys.time()
# time.taken <- end.time - start.time
# time.taken

```


### 7. Gene Ontology and Gene Data 

```{r}

# Get the GO Annotation File (GAF) format 2.1 of yeast
if (!file.exists("Data/scerevisiae_sgd.gaf.gz")){
    download.file("http://current.geneontology.org/annotations/sgd.gaf.gz", 
              destfile = "Data/scerevisiae_sgd.gaf.gz")}
if (!file.exists("Data/scerevisiae_sgd.gaf")){
    gunzip("Data/scerevisiae_sgd.gaf.gz", destname="Data/scerevisiae_sgd.gaf")}

scerevisiae_GO <- fread("Data/scerevisiae_sgd.gaf", sep = '\t', header = FALSE)
head(scerevisiae_GO)

# Subset unique GeneIDs (V11) and GO IDs (V5) in a dataframe. 
# Select the first gene synonym from column V11 and delete the rest. 
# Use this as the GeneID column. 
# Add another column with 1 in each row indicating the relationship of the Gene ID to the GO ID

scerevisiae_GO_matrix <- data.frame(unique(scerevisiae_GO[,c(11,5)]))
scerevisiae_GO_matrix$GeneID <- gsub("\\|.*", "", scerevisiae_GO_matrix$V11)
scerevisiae_GO_matrix$V3 <- as.numeric(paste(1, scerevisiae_GO_matrix$V3))
scerevisiae_GO_matrix <- scerevisiae_GO_matrix[,-1]

# Convert the dataframe to wide format with Gene ID as column 1 and GO ID as column names 
scerevisiae_GO_matrix_wide <- scerevisiae_GO_matrix %>% pivot_wider(names_from = "V5", values_from = "V3")
scerevisiae_GO_matrix_wide <- as.data.frame(scerevisiae_GO_matrix_wide)
colnames(scerevisiae_GO_matrix_wide)[1] <- "GeneID"

# convert NA to Zero
scerevisiae_GO_matrix_wide[is.na(scerevisiae_GO_matrix_wide)] <- 0

# save the dataframe to an RDS file
saveRDS(scerevisiae_GO_matrix_wide, "scerevisiae_GO_matrix_wide.rds")
head(scerevisiae_GO_matrix_wide)

# GeneIDs that are in the gene count data but not in the GO matrix
diff_Count.GO <- setdiff(rownames(agg), scerevisiae_GO_matrix_wide$GeneID)

# GeneIDs that are in the GO matrix but not in the  gene count data
diff_GO.Count <- setdiff(scerevisiae_GO_matrix_wide$GeneID, rownames(agg))

# Remove the GeneIDs that are in the GO matrix but not in the gene count data
IndexRemoved <- which(scerevisiae_GO_matrix_wide$GeneID %in% diff_GO.Count)
GO_table <- scerevisiae_GO_matrix_wide
GO_table <- GO_table[-IndexRemoved,]
# Check if the genes were properly removed 
diff_GO.Count2 <- setdiff(GO_table$GeneID, rownames(agg))
dim(diff_GO.Count2)

# Save the new GO table
saveRDS(GO_table, "GO_table.rds")


```

### 8. Get Genes with no annotations

```{r echo=TRUE}
# Use the YeastMine plugin for R from yeastgenome.org to get the list of genes
# with uncharacterised ORF

im <-  initInterMine(listMines()["YeastMine"])

q_unchar <- setQuery(
select = c("Gene.secondaryIdentifier"),
where = setConstraints(paths = c("Gene"),
operators = c("IN"),
values = list("Uncharacterized_ORFs")
))

q_dubious <- setQuery(
select = c("Gene.secondaryIdentifier"),
where = setConstraints(paths = c("Gene"),
operators = c("IN"),
values = list("Dubious_ORFs")
))

q_verified <- setQuery(
select = c("Gene.secondaryIdentifier"),
where = setConstraints(paths = c("Gene"),
operators = c("IN"),
values = list("Verified_ORFs")
))

unchar_ORFs <- runQuery(im, q_unchar)
dubious_ORFs <- runQuery(im, q_dubious)
verified_ORFs <- runQuery(im, q_verified)

unchar_ORFs_list <- as.list(unchar_ORFs$Gene.secondaryIdentifier)
dubious_ORFs_list <- as.list(dubious_ORFs$Gene.secondaryIdentifier)
verified_ORFs_list <- as.list(verified_ORFs$Gene.secondaryIdentifier)

a <- agg[which(rownames(agg) %in% unchar_ORFs_list),] #721 vs 722 in the orig
b <- agg[which(rownames(agg) %in% dubious_ORFs_list),] #688 
c <- agg[which(rownames(agg) %in% verified_ORFs_list),] #5190 vs 5195 in the orig

# 5190 + 688 + 721 = 6599 vs 6605 total; 6 missing IDs
# missing IDs

e <- setdiff(unchar_ORFs_list, rownames(a))
f <- setdiff(verified_ORFs_list, rownames(c))

# verify with the GO terms
g <- scerevisiae_GO_matrix_wide[which(scerevisiae_GO_matrix_wide$GeneID %in% verified_ORFs_list),]

h <- setdiff(verified_ORFs_list, g$GeneID) # STA1 missing in GO table

```

### 9. Data preparation functions

```{r echo=TRUE}

# A nested list with of the genes grouped per cluster and their 
# corresponding correlation values. 
corr_allClusters <- corr_per_clust(logcounts, clusters, 100)

# Save the correlations per cluster
saveRDS(corr_allClusters, "corr_allClusters.rds")


# A list of data frames containing all GO terms associated with the genes belonging to a cluster.

GOterms_allClusters <- GO_per_cl(GO_table, clusters, 100)

# Save the GO terms grouped per cluster
saveRDS(GOterms_allClusters, "GOterms_allClusters.rds")

```


### Session Information 

```{r Session Info, echo=FALSE}

sessionInfo()

```
