---
title: "Yeast Clusters with Gene Ontology"
author: "MSoria"
date: "8/24/2020"
output: html_document
---

</br>

***

##### The following code associates annotations from Gene Ontology to different clusters from the yeast genome using data from DEE2  

***

</br>


### Prerequisite libraries
```{r Libraries, echo=TRUE}
suppressPackageStartupMessages({c(library(R.utils), library(dplyr), library(tidyr), library(data.table), library(RColorBrewer), library(gplots), library(Polychrome), library(tidyverse), library(edgeR), library(GO.db), library(gage), library(gProfileR), library(factoextra), library(NbClust), library(org.Sc.sgd.db),library(stringi))})

```
</br>

###  1. Download and unzip the Yeast data 
```{r Download Data, echo=TRUE}
# put in an if statement to check if object exists

if (!file.exists("Data/scerevisiae_se.tsv.bz2")){
    download.file("http://dee2.io/mx/scerevisiae_se.tsv.bz2", 
              destfile = "Data/scerevisiae_se.tsv.bz2")}

# fread error msg: This file is very unusual: it ends abruptly without a final newline, and also its size is a multiple of 4096 bytes. Please properly end the last row with a newline using for example 'echo >> file' to avoid this  time to copy. 

# needed to go back to bunzip2
if (!file.exists("Data/scerevisiae_se.tsv")){
    bunzip2("Data/scerevisiae_se.tsv.bz2", destname="Data/scerevisiae_se.tsv")}

scerevisiae_data <- fread("Data/scerevisiae_se.tsv", sep = '\t', header = FALSE)

head(scerevisiae_data)
```
</br>

### 2. Download, read, and unzip the QC data
```{r QC Data, echo=TRUE}
if (!file.exists("Data/scerevisiae_qc.tsv.bz2")){
    download.file("http://dee2.io/mx/scerevisiae_qc.tsv.bz2", 
              destfile = "Data/scerevisiae_qc.tsv.bz2")}

quality_data <- fread("Data/scerevisiae_qc.tsv.bz2", sep = '\t', col.names = c("Database", "Description", "Details"))

head(quality_data)
```
</br>

### 3. Filter and quantify databases according to QC_SUMMARY = PASS.  
```{r QC Data Summary, echo=TRUE}

# count the total number of QC_SUMMARY entries for checking
qc_summary <- quality_data[quality_data$Description == 'QC_SUMMARY',]
total_summary <- nrow(qc_summary)

# count the total number of databases marked as PASS
qc_pass <- filter(quality_data, grepl("PASS", Details))
total_pass <- nrow(qc_pass)

```
</br>

### 4. Select all the databses from scerevisiae_se.tsv with QC_SUMMARY == 'PASS' 
```{r Filter yeast bulk data, echo=TRUE}

# filter all the sample names with QC_SUMMARY == 'PASS' from step 3
databse_pass <- filter(quality_data, grepl("PASS", Details)) 

# convert the rows into string and store on a list to use for filtering Yeast data
databse_pass_list <- as.list(as.character(databse_pass$Database))

# Filter Yeast data using generated list
scerevisiae_pass <- scerevisiae_data[scerevisiae_data$V1 %in% databse_pass_list,] 

# Convert format from long to wide
scerevisiae_pass_wide <- scerevisiae_pass %>% pivot_wider(names_from = "V1", values_from = "V3")

# Convert tibble to data frame and assign column 1 as rowname
scerevisiae_pass_wide <- as.data.frame(scerevisiae_pass_wide)
rownames(scerevisiae_pass_wide) <- scerevisiae_pass_wide[,1]
scerevisiae_pass_wide <- scerevisiae_pass_wide[,-1]

scerevisiae_pass_wide[1:10,1:6]

```
</br>

### 5. Aggregate the SRR run data to SRX experiment 
```{r SRR to SRX Aggregation, echo=TRUE}

# Download and read Metadata summary

if (!file.exists("Data/scerevisiae_metadata.tsv.cut")){
    download.file("http://dee2.io/metadata/scerevisiae_metadata.tsv.cut", 
              destfile = "Data/scerevisiae_metadata.tsv.cut")}
    
yeast_metadata <- read.csv("Data/scerevisiae_metadata.tsv.cut", sep = '\t')

# Filter metadata to include only "passed" samples
yeast_metadata_pass <- yeast_metadata[yeast_metadata$SRR_accession %in% databse_pass_list,]

# Assign column 1 as rownames 
rownames(yeast_metadata_pass) <- yeast_metadata_pass[,1]
yeast_metadata_pass <- yeast_metadata_pass[,-1]

# Sequencing Run aggregate function
srx_agg <- function(x,counts="GeneCounts") {
    IDX=which(names(x) %in% "GeneCounts")
    mds<-x$MetadataSummary
    n=nrow(x[[IDX]])
    SRX_dat <- vapply(X=unique(mds$SRX_accession) ,function(srx) {
        srrs<-rownames(mds)[which(mds$SRX_accession %in% srx)]
        if (length(srrs)>1) {
            rowSums(x[[IDX]][,srrs])
        } else {
            x[[IDX]][,srrs]
        }
    } , numeric(n))
    rownames(SRX_dat) <- rownames(x[[IDX]])
    colnames(SRX_dat) <- unique(mds$SRX_accession)
    SRX_dat
}

# Put the filtered yeast GeneCount data and the filtered metadata dataframes into a list
yeast_count_metadata <- list(GeneCounts = scerevisiae_pass_wide, MetadataSummary = yeast_metadata_pass)

# Apply both dataframes to the function
agg <- srx_agg(yeast_count_metadata)

# Check for samples with zero total counts
colSums(agg) == 0

# Histogram showing samples (columns) counts
hist(colSums(agg), breaks = 30)

# Filtering samples with more than 1 million reads
agg_f_col <- agg[, which(colSums(agg) >= 1e6)]

# Filtering genes with more than zero counts 
agg_f_row <- agg[which(rowSums(agg) > 0), ]

# No difference between the filtered and unfiltered columns  
dim(agg_f_col)
dim(agg)

# Rows ommited for genes with more than zero counts
dim(agg_f_row)


```
</br>

### 6. Normalisation of Data 

```{r eval=FALSE, include=FALSE}
# NORMALISATION FOR LIBRARY SIZE BIAS (cpm)
myCPM <- cpm(agg)
# add a small number to the entire dataframe to account for zero reads
myCPM <- myCPM + 0.01
# Filtering to remove lowly expressed genes
thresh <- myCPM > 0.6
# check if there are zero values. Result should be FALSE.
all(apply(apply(myCPM, 2, function(x) x==0), 2, any))

# Keep genes that have at least 2 TRUEs in each row of thresh
keep <- rowSums(thresh) >= 2

# Subset the rows of data to keep the more highly expressed genes
counts_keep <- myCPM[keep,]
summary(keep)
dim(counts_keep)

# A CPM of 0.6 is used as it corresponds to a count of 10-15 for the library sizes in this data set. If the count is any smaller, it is considered to be very low, indicating that the associated gene is not expressed in that sample. As a general rule, a good threshold can be chosen by identifying the CPM that corresponds to a count of 10, which in this case is about 0.6. You should filter with CPMs rather than filtering on the counts directly, as the latter does not account for differences in library sizes between samples.

# Note: When in doubt, a threshold of 1 CPM in at least minimum group sample size is a good rule of thumb.

# Check whether our threshold of 0.6 does indeed correspond to a count of about 10-15
# Add a vertical line at 0.6 CPM, and a horizontal line at Counts = 10
plot(myCPM[,1], agg[,1], ylim=c(0,50), xlim=c(0,3), abline(v=0.6, h=10, col="blue"), ylab = "Counts", xlab = "CPM", main = "Counts-per-million (CPM) vs Counts")

# Convert counts to DGEList object (from edgeR library)
dgeObj <- DGEList(counts_keep)

#NORMALISATION FOR COMPOSITION BIAS (TMM normalisation)
# Apply normalisation to DGEList object
dgeObj <- calcNormFactors(dgeObj) 
# dgeObj$samples

# Get log2 counts per million
logcounts <- cpm(dgeObj,log=TRUE)

# Check distributions of samples
hist(logcounts)
hist(counts_keep)

all(apply(apply(logcounts, 2, function(x) x==0), 2, any))
# Comment: should we remove the huge spike of less than 5 values?

```
</br>


### 8. Heirarchical Clustering with PASS only databases
```{r echo=TRUE}

# Applying Hierarchical Clustering after normalization
cl <-as.dist(1-cor(t(logcounts), method="spearman"))
hr <- hclust(cl , method="complete")

# optimizing the cluster size
mycl <- cutree(hr, h=max(hr$height/1.64))
mycl_length <- length(unique(mycl))

if (!exists("P100")){
  P100 <- createPalette(100,  c("#ff0000", "#00ff00", "#0000ff"))}

clusterCols <- P100[1:mycl_length]
myClusterSideBar <- clusterCols[mycl]
colfunc <- colorRampPalette(c("blue", "white", "red"))
if (!file.exists("Data/scerevisiae_se.tsv")){
  write.table(mycl,file="Data/scerevisiae_pass.txt",quote=F,sep="\t")}

# create a 5 x 5 inch png image
png("Data/yeast_heatmaps_PASS_normalization3.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)          

# create the heatmap
heatmap.2(logcounts, main="S.cerevisiae PASS Samples",  Rowv=as.dendrogram(hr),
          dendrogram="both", scale="column", col = colfunc(25), trace="none", 
          RowSideColors= myClusterSideBar, margins = c(5,5))

dev.off()
```
</br>



### 9. Gene Ontology and Gene Data 
```{r}
# Get the GO Annotation File (GAF) format 2.1 of yeast
if (!file.exists("Data/scerevisiae_sgd.gaf.gz")){
    download.file("http://current.geneontology.org/annotations/sgd.gaf.gz", 
              destfile = "Data/scerevisiae_sgd.gaf.gz")}
if (!file.exists("Data/scerevisiae_sgd.gaf")){
    gunzip("Data/scerevisiae_sgd.gaf.gz", destname="Data/scerevisiae_sgd.gaf")}

scerevisiae_GO <- fread("Data/scerevisiae_sgd.gaf", sep = '\t', header = FALSE)
head(scerevisiae_GO)

# Subset unique GeneIDs (V11) and GO IDs (V5) in a dataframe. Select the first gene synonym from column V11 and delete the rest. Use this as the GeneID column. Add another column with 1 in each row indicating the relationship of the Gene ID to the GO ID

scerevisiae_GO_matrix <- data.frame(unique(scerevisiae_GO[,c(11,5)]))
scerevisiae_GO_matrix$GeneID <- gsub("\\|.*", " ", scerevisiae_GO_matrix$V11)
scerevisiae_GO_matrix$V3 <- as.numeric(paste(1, scerevisiae_GO_matrix$V3))
scerevisiae_GO_matrix <- scerevisiae_GO_matrix[,-1]

# Convert the dataframe to wide format with Gene ID as column 1 and GO ID as column names 
scerevisiae_GO_matrix_wide <- scerevisiae_GO_matrix %>% pivot_wider(names_from = "V5", values_from = "V3")
scerevisiae_GO_matrix_wide <- as.data.frame(scerevisiae_GO_matrix_wide)
colnames(scerevisiae_GO_matrix_wide)[1] <- "GeneID"

# convert NA to Zero
scerevisiae_GO_matrix_wide[is.na(scerevisiae_GO_matrix_wide)] <- 0

# save the dataframe to an RDS file
saveRDS(scerevisiae_GO_matrix_wide, "scerevisiae_GO_matrix_wide.rds")
head(scerevisiae_GO_matrix_wide)

```
</br>


### 10. Blinding genes using random strings
```{r}
#Fraction of the total number of genes to be blinded
frac_blind <- 0.10
number_of_genes <- round(nrow(agg)*frac_blind)

blind <- sample(rownames(agg), number_of_genes)

rand <- stri_rand_strings(number_of_genes, 15, pattern = "[A-Za-z]")
rand <- paste("blind_", rand, sep = "")
dict_blind <- data.frame(blind, rand)

# save the dictionary into an RDS file
saveRDS(dict_blind, "dict_blind.rds")

agg_rand <- agg

ind_blind <- which(rownames(agg_rand) %in% blind)

rownames(agg_rand)[ind_blind] <- rand

```
</br>


### 11. Functions for Cluster Analysis 
```{r echo=TRUE}

# This function's output is a nested list with of the genes grouped per cluster and their corresponding correlation values. Inputs:
# x = logcounts (normalized gene counts from RNA seq)
# y = clusters (matrix of genes w/ cluster number)
# clust_total = total number of clusters

corr_per_clust <- function(x, y, clust_total){
  corr_cl <- list()
  for (i in 1:clust_total){
    clusterX <- y[y$ClusterNumber == i,]
    cluster_list <- as.list(clusterX$GeneID)
    cluster <- x[rownames(x) %in% cluster_list,]
    corr_result <- cor(t(cluster))
    corr_cl[[paste0("Cluster", i)]] <- corr_result
  }
  return(corr_cl)
}

corr_allClusters <- corr_per_clust(logcounts, clusters, 100)


# This function's output is a list of data frames containing all GO terms associated with the genes belonging to a cluster. Input:
# x = scerevisiae_GO_matrix_wide (matrix of genes belonging to a GO term)
# y = clusters (matrix of genes w/ cluster number)
# clust_total = total number of clusters

GO_per_cl <- function(x,y,clust_total){
  GO_cl <- list()
  for (i in 1:clust_total){
    clusterX <- y[y$ClusterNumber == i,]
    cluster_list <- as.list(clusterX$GeneID)
    cluster_GOterms <- x[x$GeneID %in% cluster_list,]
    rownames(cluster_GOterms)<- cluster_GOterms[,1] 
    cluster_GOterms[,1] <- c()
    cluster_GOterms <- cluster_GOterms[,which(colSums(cluster_GOterms) > 0)]
    GO_cl[[paste0("Cluster", i)]] <- cluster_GOterms
  }
  return(GO_cl)
}

GOterms_allClusters <- GO_per_cl(scerevisiae_GO_matrix_wide, clusters, 100)


# This function's output is a nested list of genes belonging to a cluster with their corresponding weighted correlations. Input:
# gene_list = list of genes to be correlated; corr_cl = correlation values for cluster
# cl_GO = GO terms for cluster

wcorr_cluster <- function(gene_list, corr_cl, cl_GO){
  wcorr_result <- list()
  
  for (i in gene_list){
    corr_value <- corr_cl[i,]
    corr_value_df <- as.data.frame(corr_value)
    weighted_go <- merge(corr_value_df, cl_GO, by = "row.names")
    rownames(weighted_go) <- weighted_go[,1]
    weighted_go[,1] <- c()
    weighted_go <- weighted_go[,1]*weighted_go[,2:ncol(weighted_go)]
    normalized_weighted_go <- colSums(weighted_go)/nrow(weighted_go)
    wcorr_result[[i]] <- normalized_weighted_go
  }
  setNames(wcorr_result, paste0(i))
  return(wcorr_result)
}

cluster1_wGO <- wcorr_cluster(is_blind, corr_cluster1, cluster1_GO_terms)


# Sum of corr values are used as a denominator in normalized_weighted_go
wcorr2_cluster <- function(gene_list, corr_cl, cl_GO){
  wcorr_result <- list()
  
  for (i in gene_list){
    corr_value <- corr_cl[i,]
    corr_value_df <- as.data.frame(corr_value)
    weighted_go <- merge(corr_value_df, cl_GO, by = "row.names")
    rownames(weighted_go) <- weighted_go[,1]
    weighted_go[,1] <- c()
    # this is where this differs from the original function. The sum of the correlation values for         gene1 in the gene_list is used as the denominator for the normalized_weighted_go object instead of     the nrow of the dataframe. This follows the formula of the weighted average, assuming that the corr     values are the "weights" 
    corr_colsum <- sum(weighted_go[,1])
    weighted_go <- weighted_go[,1]*weighted_go[,2:ncol(weighted_go)]
    normalized_weighted_go <- colSums(weighted_go)/corr_colsum
    wcorr_result[[i]] <- normalized_weighted_go
  }
  setNames(wcorr_result, paste0(i))
  return(wcorr_result)
}

cluster1_wGO2 <- wcorr2_cluster(is_blind, corr_cluster1, cluster1_GO_terms)




```
</br>

### 12. Cluster Analysis
```{r}
# Prepare Cluster data frame
clusters <- as.data.frame(mycl)
colnames(clusters) <- "ClusterNumber"

# make a new column, GeneID, from rownames
clusters$GeneID <- rownames(clusters)

# Grouping genes by cluster (using cluster 1)
cluster1 <- clusters[clusters$ClusterNumber == 1,] 
cluster1_list <- cluster1$GeneID

# look for cluster gene match in GO dictionary and subset nonzero columns 
cluster1_GO <- scerevisiae_GO_matrix_wide[scerevisiae_GO_matrix_wide$GeneID %in% cluster1_list,] 
cluster1_GO_nonzero <- cluster1_GO[, !apply(cluster1_GO == 0, 2, all)]
cluster1_GO_nonzero

# Correlation in Cluster 1  (Function counterpart: corr_per_clust)
cluster_1 <- logcounts[rownames(logcounts) %in% cluster1_list,]
corr_cluster1 <- cor(t(cluster_1))

corr2_cluster1 <- cor(cluster_1)

# GO terms for cluster 1 (Function counterpart: GO_per_cl)
cluster1_GO_terms <- scerevisiae_GO_matrix_wide[scerevisiae_GO_matrix_wide$GeneID %in% cluster1_list,]
rownames(cluster1_GO_terms)<- cluster1_GO_terms[,1] 
cluster1_GO_terms[,1] <- c()
cluster1_GO_terms <- cluster1_GO_terms[,which(colSums(cluster1_GO_terms) > 0)]

# weighted correlation for Cluster 1 (Function counterpart: wcorr_cluster)
is_blind <- cluster1_list[cluster1_list %in% dict_blind$blind]
blind_gene1 <- is_blind[1]

# Subset the row which contains blilnd_gene1 and turn it into a data frame
corr_value_blind1 <- corr_cluster1[blind_gene1,]
corr_value_blind1_df <- as.data.frame(corr_value_blind1)

# Merge the GO terms data frame for cluster 1 and the correlation values for blilnd_gene1 
weighted_go_cluster1 <- merge(corr_value_blind1_df, cluster1_GO_terms, by = "row.names")
rownames(weighted_go_cluster1) <- weighted_go_cluster1[,1]
weighted_go_cluster1[,1] <- c()
corr_colsum <- sum(weighted_go_cluster1[,1])
weighted_go_cluster1 <- weighted_go_cluster1[,1]*weighted_go_cluster1[,2:ncol(weighted_go_cluster1)]
normalized_weighted_go_cluster1 <- colSums(weighted_go_cluster1)/corr_colsum 
hist(normalized_weighted_go_cluster1)


# Use function, wcorr_cluster, to get the weighted values for all blinded genes in cluster 1
cluster1_wGO <- wcorr_cluster(is_blind, corr_cluster1, cluster1_GO_terms)

# Determining threshold based on histogram
hist(cluster1_wGO$YPR002W)
imputed_GO_terms <- names(which(normalized_weighted_go_cluster1 > 0.005))

# Convert output into a data frame with columns as GO terms and rows as gene names
cluster1_wGO_df <- as.data.frame(do.call(rbind, cluster1_wGO))
cluster1_wGO_df <- cluster1_wGO_df[order(rownames(cluster1_wGO_df)),]
cluster1_wGO_df <- cluster1_wGO_df[,order(colnames(cluster1_wGO_df))]

# using 0.1 as threshold
Clust1_threshold <- 0.1

imputed_Clust1_df_v2 <- (as.matrix(cluster1_wGO_df) > Clust1_threshold)*1




# Filter original GO matrix to include only the genes of interest and the GO terms from the imputed df
# Filter by row
cluster1_inputMat <- scerevisiae_GO_matrix_wide[scerevisiae_GO_matrix_wide$GeneID %in% rownames(imputed_Clust1_df),]
rownames(cluster1_inputMat) <- cluster1_inputMat[,1]
cluster1_inputMat[,1] = c()
# Filter by column
cluster1_inputMat <- cluster1_inputMat[,colnames(cluster1_inputMat) %in% colnames(imputed_Clust1_df)]
# Order by row names and column names
cluster1_inputMat1 <- cluster1_inputMat[order(rownames(cluster1_inputMat)),]
cluster1_inputMat1 <- cluster1_inputMat[,order(colnames(cluster1_inputMat))]


# Filter imputed_Clust1_df object to include the genes and GO terms of cluster1_inputMat object above
imputed_Clust1_trim <- imputed_Clust1_df[rownames(imputed_Clust1_df_v2) %in% rownames(cluster1_inputMat),]
# Order by row names for direct comparison
imputed_Clust1_trim <- imputed_Clust1_trim[order(rownames(imputed_Clust1_trim)),]


# Comparison between input df and resultant df
Clust1_inutVSimputed <- ifelse(cluster1_inputMat == imputed_Clust1_trim, TRUE, FALSE)

# Negative 1 values appear for threshold > 0.005
Clust1_subtract <- imputed_Clust1_trim - cluster1_inputMat
which(Clust1_subtract == -1)

```
</br>

### Using wcorr2_cluster
```{r echo=TRUE}

cluster1_wGO2 <- wcorr2_cluster(is_blind, corr_cluster1, cluster1_GO_terms)

# Convert output into a data frame with columns as GO terms and rows as gene names
cluster1_wGO2_df <- as.data.frame(do.call(rbind, cluster1_wGO2))
cluster1_wGO2_df <- cluster1_wGO2_df[order(rownames(cluster1_wGO2_df)),]
cluster1_wGO2_df <- cluster1_wGO2_df[,order(colnames(cluster1_wGO2_df))]

# using 0.01 as threshold
Clust1_threshold <- 0.01

imputed_Clust1_df <- as.data.frame(sapply(cluster1_wGO2_df, function(x){
   as.numeric(x > Clust1_threshold)
  }))
row.names(imputed_Clust1_df) <- row.names(cluster1_wGO2_df)

# Filter original GO matrix to include only the genes of interest and the GO terms from the imputed df
# Filter by row
cluster1_inputMat <- scerevisiae_GO_matrix_wide[scerevisiae_GO_matrix_wide$GeneID %in% rownames(imputed_Clust1_df),]
rownames(cluster1_inputMat) <- cluster1_inputMat[,1]
cluster1_inputMat[,1] = c()
# Filter by column
cluster1_inputMat <- cluster1_inputMat[,colnames(cluster1_inputMat) %in% colnames(imputed_Clust1_df)]
# Order by row names and column names
cluster1_inputMat1 <- cluster1_inputMat[order(rownames(cluster1_inputMat)),]
cluster1_inputMat1 <- cluster1_inputMat[,order(colnames(cluster1_inputMat))]


# Filter imputed_Clust1_df object to include the genes and GO terms of cluster1_inputMat object above
imputed_Clust1_trim <- imputed_Clust1_df[rownames(imputed_Clust1_df) %in% rownames(cluster1_inputMat),]
# Order by row names for direct comparison
imputed_Clust1_trim <- imputed_Clust1_trim[order(rownames(imputed_Clust1_trim)),]
imputed_Clust1_trim <- imputed_Clust1_trim[,order(colnames(imputed_Clust1_trim))]

# Comparison between input df and resultant df
Clust1_inutVSimputed <- ifelse(cluster1_inputMat == imputed_Clust1_trim, TRUE, FALSE)

# Negative 1 values appear for threshold > 0.01
Clust1_subtract <- imputed_Clust1_trim - cluster1_inputMat
which(Clust1_subtract == -1)

```
</br>

### Using cluster 3
```{r echo=TRUE}

corr_cluster3 <- corr_allClusters$Cluster3
cluster3_GO_terms <- GOterms_allClusters$Cluster3
cluster3_list <- rownames(cluster3_GO_terms)
# Make a function to list all blind genes in a cluster
clust3_blind <- cluster3_list[cluster3_list %in% dict_blind$blind]

cluster3_wGO2 <- wcorr2_cluster(clust3_blind, corr_cluster3, cluster3_GO_terms)

# Convert output into a data frame with columns as GO terms and rows as gene names
cluster3_wGO2_df <- as.data.frame(do.call(rbind, cluster3_wGO2))
cluster3_wGO2_df <- cluster3_wGO2_df[order(rownames(cluster3_wGO2_df)),]
cluster3_wGO2_df <- cluster3_wGO2_df[,order(colnames(cluster3_wGO2_df))]

# using 0.01 as threshold
Clust3_threshold <- 0.01

imputed_Clust3_df <- as.data.frame(sapply(cluster3_wGO2_df, function(x){
   as.numeric(x > Clust3_threshold)
  }))
imputed_Clust3_df <- t(imputed_Clust3_df)
row.names(imputed_Clust3_df) <- row.names(cluster3_wGO2_df)

# Filter original GO matrix to include only the genes of interest and the GO terms from the imputed df
# Filter by row
cluster3_inputMat <- scerevisiae_GO_matrix_wide[scerevisiae_GO_matrix_wide$GeneID %in% rownames(imputed_Clust3_df),]
rownames(cluster3_inputMat) <- cluster3_inputMat[,1]
cluster3_inputMat[,1] = c()
# Filter by column
cluster3_inputMat <- cluster3_inputMat[,colnames(cluster3_inputMat) %in% colnames(imputed_Clust3_df)]
# Order by row names and column names
cluster3_inputMat <- cluster3_inputMat[order(rownames(cluster3_inputMat)),]
cluster3_inputMat <- cluster3_inputMat[,order(colnames(cluster3_inputMat))]


# Filter imputed_Clust3_df object to include the genes and GO terms of cluster1_inputMat object above
imputed_Clust3_trim <- imputed_Clust3_df[rownames(imputed_Clust3_df) %in% rownames(cluster3_inputMat),]

# Comparison between input df and resultant df
Clust3_inutVSimputed <- ifelse(cluster3_inputMat == imputed_Clust3_trim, TRUE, FALSE)

# Negative 1 values appear for threshold > 0.01
Clust3_subtract <- imputed_Clust3_trim - cluster3_inputMat
which(Clust3_subtract == -1)
```

</br>


### 13. Gene Enrichment Analysis 
```{r echo=TRUE}
# ---------Functional Enrichment Analysis using using goProfileR

# Gene enrichment query
goResults_1 <- gprofiler(query = genesOfInterest,
                     organism = 'scerevisiae',
                     hier_filtering = 'none')

## Subset and reorder gProfiler results to only include columns of interest
gprofiler_results_oe_reordered <- goResults_1[, c("term.id", "domain", "term.name", "p.value", "overlap.size", "term.size", "intersection")]

# Order the results by p-adjusted value
gprofiler_results_oe_reordered <- gprofiler_results_oe_reordered[order(gprofiler_results_oe_reordered$p.value), ]
 
# Extract only the 'GO' terms from the results
gprofiler_results_oe_GOs <- gprofiler_results_oe_reordered[grep('GO:', gprofiler_results_oe_reordered$term.id), ]

# Write the enriched GO results to file
write.csv(gprofiler_results_oe_GOs,
            "Data/gprofiler_cluster1.csv")


# --------Over-representation analysis of biological processes
# perform Gene Set Enrichment Analysis (GSEA) using gage package

columns(org.Sc.sgd.db)
keys(org.Sc.sgd.db)
keytypes(org.Sc.sgd.db)

# The input to gage must be Entrez identifiers so we first need to map our gene symbols to Entrez
ann <- mapIds(org.Sc.sgd.db, keys=rownames(cluster_1), column="ENTREZID", keytype="ENSEMBL")
cluster1_non_na <- names(!is.na(ann))
fc <- logcounts[rownames(logcounts) %in% cluster1_non_na,]

table(is.na(fc))

# Create a mapping between gene identifiers and GO terms.
allGo <- as.list(GOTERM)
bp <- Filter(function(go) go@Ontology == 'BP', allGo)
goMapping <- mget(names(bp), org.Sc.sgdGO2ALLORFS, ifnotfound=NA)

# gene set enrichment analysis
goFc <- gage(fc, goMapping)
goGreater <- as.data.frame(goFc$greater)
goGreater <- goGreater[goGreater$q.val < 0.01 & !is.na(goGreater$q.val),]
goTermNames <- lapply(mget(rownames(goGreater), GOTERM), function(go) go@Term)
goGreater$Term <- goTermNames[rownames(goGreater)]
goGreater



#---------using clusterprofiler
library("clusterProfiler")


background_genes <- clusters[clusters$ClusterNumber !=1,]
background_geneIDs <- rownames(agg)

yeast_CP_go <- enrichGO(genesOfInterest, "org.Sc.sgd.db", keyType = "ENSEMBL", ont = "ALL", universe = background_geneIDs)
yeast_CP_go_df <- as.data.frame(yeast_CP_go)
yeast_CP_go_df
```

</br>


### 14. Assessing clustering tendency

```{r echo=TRUE}

# Hopkins statistic: If the value of Hopkins statistic is close to 1 (far above 0.5), 
# then we can conclude that the dataset is significantly clusterable

# Visual approach: The visual approach detects the clustering tendency by counting 
# the number of square shaped dark (or colored) blocks along the diagonal in 
# the ordered dissimilarity image.

gradient.color <- list(low = "steelblue",  high = "white")

# create a 5 x 5 inch png image
png("Data/clustering_tendency_PASS.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)          
logcounts %>% get_clust_tendency(n = nrow(logcounts)-1, gradient = gradient.color)

dev.off()
```
</br>

### 15. K-means Clustering

```{r echo=TRUE}
# determining the number of clusters (K) using NbClust package

#val_K_clust <- NbClust(logcounts, min.nc=1, max.nc=100, method="kmeans")
num_K_clust <- fviz_nbclust(logcounts, FUNcluster = kmeans, method = "wss", print.summary = TRUE)

# K-means Clustering

# As k-means clustering algorithm starts with k randomly selected centroids, it’s always recommended to use the set.seed() function in order to set a seed for R’s random number generator

set.seed(123)
km.res_100 <- kmeans(logcounts, 100, nstart = 25, iter.max = 50)

png("Data/K_means_PASS.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)
fviz_cluster(km.res_100, data = logcounts, geom="point")
dev.off()
```
</br>

### 16. Applying Function for K-means results
```{r echo=TRUE}

```


</br>

### Session Information 

```{r Session Info, echo=FALSE}
sessionInfo()
```