---
title: "Yeast Sample Quality Filter "
author: "Megan Soria"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
theme: cosmo
---
</br>

***

##### The following code will determine the quality of the runs for S.Cerevisiae (Yeast) RNA-seq data downloaded from the [Digital Expression Explorer 2](http://dee2.io/) repository. It will quantify the number of PASS, WARN and FAIL runs and will filter samples with PASS runs  

***

</br>


### Prerequisite libraries
```{r Libraries, echo=TRUE}
suppressPackageStartupMessages({c(library(R.utils), library(dplyr), library(tidyr), library(data.table), library(RColorBrewer), library(gplots), library(viridis), library(DESeq2))})

```
</br>

###  1. Download and unzip the Yeast data 
```{r Download Data, echo=TRUE}
# put in an if statement to check if object exists

download.file("http://dee2.io/mx/scerevisiae_se.tsv.bz2", 
              destfile = "Data/scerevisiae_se.tsv.bz2")

# fread error msg: This file is very unusual: it ends abruptly without a final newline, and also its size is a multiple of 4096 bytes. Please properly end the last row with a newline using for example 'echo >> file' to avoid this  time to copy. 

# needed to go back to bunzip2

bunzip2("Data/scerevisiae_se.tsv.bz2", destname="Data/scerevisiae_se.tsv")

scerevisiae_data <- fread("Data/scerevisiae_se.tsv", sep = '\t', header = FALSE)

head(scerevisiae_data)
```
</br>

### 2. Download, read, and unzip the QC data
```{r QC Data, echo=TRUE}
download.file("http://dee2.io/mx/scerevisiae_qc.tsv.bz2", 
              destfile = "Data/scerevisiae_qc.tsv.bz2")

quality_data <- fread("Data/scerevisiae_qc.tsv.bz2", sep = '\t', col.names = c("Database", "Description", "Details"))

head(quality_data)
```
</br>

### 3. Filter and quantify databases according to QC_SUMMARY = PASS, WARN, or FAIL.  
```{r QC Data Summary, echo=TRUE}

# count the total number of QC_SUMMARY entries for checking
qc_summary <- quality_data[quality_data$Description == 'QC_SUMMARY',]
total_summary <- nrow(qc_summary)

# count the total number of databases marked as PASS
qc_pass <- filter(quality_data, grepl("PASS", Details))
total_pass <- nrow(qc_pass)

# count the total number of databases marked as WARN
qc_warn <- filter(quality_data, grepl("WARN", Details))
total_warn <- nrow(qc_warn)

# count the total number of databases marked as FAIL
qc_fail <- filter(quality_data, grepl("FAIL", Details))
total_fail <- nrow(qc_fail)

# Summary of counts. Addition of each PASS, WARN, and FAIL counts equals to total_summary.
quality_summary <- data.frame(total_pass, total_warn, total_fail, total_summary)
quality_summary

```
</br>

### 4. Select all the databses from scerevisiae_se.tsv with QC_SUMMARY == 'PASS' 
```{r Filter yeast bulk data, echo=TRUE}

# filter all the sample names with QC_SUMMARY == 'PASS' from step 3
databse_pass <- filter(quality_data, grepl("PASS", Details)) %>% select(Database)

# convert the rows into string and store on a list to use for filtering Yeast data
databse_pass_list <- as.list(as.character(databse_pass$Database))

# Filter Yeast data using generated list
scerevisiae_pass <- scerevisiae_data[scerevisiae_data$V1 %in% databse_pass_list,] 

# Convert format from long to wide
scerevisiae_pass_wide <- scerevisiae_pass %>% pivot_wider(names_from = "V1", values_from = "V3")

# Convert tibble to data frame and assign column 1 as rowname
scerevisiae_pass_wide <- as.data.frame(scerevisiae_pass_wide)
rownames(scerevisiae_pass_wide) <- scerevisiae_pass_wide[,1]
scerevisiae_pass_wide <- scerevisiae_pass_wide[,-1]

scerevisiae_pass_wide[1:10,1:6]

```
</br>

### 5. Aggregate the SRR run data to SRX experiment 
```{r SRR to SRX Aggregation, echo=TRUE}

# Download and read Metadata summary
download.file("http://dee2.io/metadata/scerevisiae_metadata.tsv.cut", 
              destfile = "Data/scerevisiae_metadata.tsv.cut")
yeast_metadata <- read.csv("Data/scerevisiae_metadata.tsv.cut", sep = '\t')

# Filter metadata to include only "passed" samples
yeast_metadata_pass <- yeast_metadata[yeast_metadata$SRR_accession %in% databse_pass_list,]

# Assign column 1 as rownames 
rownames(yeast_metadata_pass) <- yeast_metadata_pass[,1]
yeast_metadata_pass <- yeast_metadata_pass[,-1]

# Sequencing Run aggregate function
srx_agg <- function(x,counts="GeneCounts") {
    IDX=which(names(x) %in% "GeneCounts")
    mds<-x$MetadataSummary
    n=nrow(x[[IDX]])
    SRX_dat <- vapply(X=unique(mds$SRX_accession) ,function(srx) {
        srrs<-rownames(mds)[which(mds$SRX_accession %in% srx)]
        if (length(srrs)>1) {
            rowSums(x[[IDX]][,srrs])
        } else {
            x[[IDX]][,srrs]
        }
    } , numeric(n))
    rownames(SRX_dat) <- rownames(x[[IDX]])
    colnames(SRX_dat) <- unique(mds$SRX_accession)
    SRX_dat
}

# Put the filtered yeast GeneCount data and the filtered metadata dataframes into a list
yeast_count_metadata <- list(GeneCounts = scerevisiae_pass_wide, MetadataSummary = yeast_metadata_pass)

# Apply both dataframes to the function
agg <- srx_agg(yeast_count_metadata)

agg[1:10,1:6]

# Check for samples with zero total counts
colSums(agg) == 0

# Histogram showing samples (columns) counts
hist(colSums(agg), breaks = 30)

# Filtering samples with more than 1 million reads
agg_f_col <- agg[, which(colSums(agg) >= 1e6)]


# No difference between the filtered and unfiltered columns and 
dim(agg_f_col_)
dim(agg)

# Filtering genes with more than zero counts 
agg_f_row <- agg[which(rowSums(agg) > 0), ]
dim(agg_f_row)


```
</br>

### 6. Heirarchical Clustering with PASS only databases 

```{r eval=FALSE, include=FALSE}



# Cluster the filtered genes with > 0 total counts
cl <-as.dist(1-cor(t(yeast_f_row), method="spearman"))
hr <- hclust(cl , method="complete")

# optimizing the cluster size
mycl <- cutree(hr, h=max(hr$height/1.3))
clusterCols <- brewer.pal(length(unique(mycl)),"Paired")
myClusterSideBar <- clusterCols[mycl]
colfunc <- colorRampPalette(c("blue", "white", "red"))
write.table(mycl,file="Data/scerevisiae_pass.txt",quote=F,sep="\t")

# create a 5 x 5 inch png image
png("Data/yeast_heatmaps_pass_yeast_row.png", width = 5*300, height = 5*300, res = 300, pointsize = 8)               

# create the heatmap
heatmap.2(yeast_f_row, main="S.cerevisiae PASS Samples",  Rowv=as.dendrogram(hr),
          dendrogram="both", scale="none", col = colfunc(25), trace="none", 
          RowSideColors= myClusterSideBar, margins = c(5,5))

dev.off()
```


</br>

### 7. Heirarchical Clustering with PASS and WARN databases 
```{r echo=TRUE}

```

### random stings

```{r}
library("stringi")

blind <- sample(rownames(agg_f_row), 10)

rand <- stri_rand_strings(10, 15, pattern = "[A-Za-z]")


dict <- data.frame(blind, rand)

agg_rand <- agg_f_row

ind_blind <- which(rownames(agg_rand) %in% blind)

rownames(agg_rand)[ind_blind] <- rand


```


### Session Information 

```{r Session Info, echo=FALSE}
sessionInfo()
```


